File: backend/utils/html_sanitizer.py

from __future__ import annotations

import re

CSP_META = (
    '<meta http-equiv="Content-Security-Policy" '
    'content="default-src \'none\'; style-src \'unsafe-inline\'; '
    'img-src data: file:; font-src data:;">'
)

SAFE_DATA_URI_PREFIXES = (
    "data:image/png",
    "data:image/jpeg",
    "data:image/jpg",
    "data:image/gif",
    "data:image/bmp",
    "data:image/webp",
    "data:image/x-icon",
)


# Numeric refs decode in browsers without the trailing semicolon; named refs
# keep it required (the no-semicolon legacy named set is not decoded inside
# attributes when followed by alnum/'=', so requiring ';' matches that).
_URL_ENTITY_RE = re.compile(r"&(?:#(x[0-9a-fA-F]+|\d+);?|([a-zA-Z][a-zA-Z0-9]*);)")
_NAMED_URL_ENTITIES = {
    "colon": ":", "tab": "\t", "newline": "\n", "sol": "/", "bsol": "\\",
    "period": ".", "amp": "&", "lt": "<", "gt": ">", "quot": '"', "apos": "'",
}


def is_safe_data_url(url: str) -> bool:
    lowered = str(url).strip().lower()
    return any(lowered.startswith(prefix) for prefix in SAFE_DATA_URI_PREFIXES)


def _decode_url_entities(value: str) -> str:
    def _sub(match: re.Match[str]) -> str:
        num, name = match.group(1), match.group(2)
        if num is not None:
            try:
                code = int(num[1:], 16) if num[0].lower() == "x" else int(num, 10)
            except ValueError:
                return match.group(0)
            return chr(code) if 0 <= code <= 0x10FFFF else match.group(0)
        mapped = _NAMED_URL_ENTITIES.get(name.lower())
        return mapped if mapped is not None else match.group(0)

    return _URL_ENTITY_RE.sub(_sub, str(value))


def _collapse_unsafe_url(match: re.Match[str]) -> str:
    url_value = _decode_url_entities(match.group(2))
    if is_safe_data_url(url_value):
        return match.group(0)
    return "url('')"


def _neutralize_url_attr(match: re.Match[str]) -> str:
    attr = match.group(1)
    quote = match.group(2) or ""
    url_value = _decode_url_entities(match.group(3))
    import urllib.parse

    cleaned = re.sub(r"\s+", "", url_value.strip()).lower()
    cleaned = re.sub(r"\s*:\s*", ":", cleaned)
    try:
        parsed = urllib.parse.urlparse(cleaned)
        scheme = parsed.scheme.lower()
    except Exception:
        scheme = ""
    if cleaned.startswith("data:"):
        if not is_safe_data_url(url_value):
            return f"{attr}={quote}{quote}"
        if cleaned.startswith("data:text/html"):
            return f"{attr}={quote}{quote}"
        return match.group(0)
    if scheme in ("javascript", "vbscript") or cleaned.startswith(("javascript:", "vbscript:")):
        return f"{attr}={quote}{quote}"
    if scheme in ("http", "https", "file"):
        return f"{attr}={quote}{quote}"
    if any(cleaned.startswith(p) for p in ("javascript:", "vbscript:", "http:", "https:", "file:")):
        return f"{attr}={quote}{quote}"
    return match.group(0)


def _neutralize_srcset_attr(match: re.Match[str]) -> str:
    attr = match.group(1)
    quote = match.group(2) or ""
    decoded = _decode_url_entities(match.group(3))
    for candidate in decoded.split(","):
        parts = candidate.split()
        url = parts[0] if parts else ""
        cleaned = re.sub(r"\s+", "", url).lower()
        if cleaned.startswith("data:"):
            if not is_safe_data_url(url) or cleaned.startswith("data:text/html"):
                return f"{attr}={quote}{quote}"
            continue
        scheme_match = re.match(r"^([a-z][a-z0-9+.-]*):", cleaned)
        scheme = scheme_match.group(1) if scheme_match else ""
        if scheme in ("javascript", "vbscript", "http", "https", "file"):
            return f"{attr}={quote}{quote}"
    return match.group(0)


def _dispatch_url_attr(match: re.Match[str]) -> str:
    if str(match.group(1)).lower() == "srcset":
        return _neutralize_srcset_attr(match)
    return _neutralize_url_attr(match)


def _neutralize_import_statement(match: re.Match[str]) -> str:
    url_value = _decode_url_entities(match.group(2))
    if is_safe_data_url(url_value):
        return match.group(0)
    return ""


def sanitize_html_for_pdf(html: str) -> str:
    stripped = str(html)
    stripped = re.sub(r"<!--[\s\S]*?-->", "", stripped)
    stripped = re.sub(r'<meta[^>]+http-equiv=["\']?Content-Security-Policy["\']?[^>]*>', "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<meta[^>]*http-equiv[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<base[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"</base>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<script[^>]*>[\s\S]*?</script>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<iframe[^>]*>[\s\S]*?</iframe>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<object[^>]*>[\s\S]*?</object>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<embed[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<link[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<script[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"</script>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"<iframe[^>]*>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"</iframe>", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r'\son[a-z]+\s*=\s*"[^"]*"', "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"\son[a-z]+\s*=\s*'[^']*'", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"\son[a-z]+\s*=\s*`[^`]*`", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"\son[a-z]+\s*=\s*[^\s>]+", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(r"\son[a-z]+\b(?=\s|>|/)", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(
        r'(href|src|xlink:href|srcset|poster|action|formaction|background|cite|ping|longdesc)\s*=\s*(["\']?)\s*([^"\'>]+)\2',
        _dispatch_url_attr,
        stripped,
        flags=re.IGNORECASE,
    )
    stripped = re.sub(
        r'@import\s+(["\'])([^"\']*)\1\s*;?',
        _neutralize_import_statement,
        stripped,
        flags=re.IGNORECASE,
    )
    stripped = re.sub(r"expression\s*\(", "", stripped, flags=re.IGNORECASE)
    stripped = re.sub(
        r"url\(\s*(['\"]?)\s*(?:javascript|vbscript):[^'\")\s]*\1\s*\)",
        "url('')",
        stripped,
        flags=re.IGNORECASE,
    )
    stripped = re.sub(
        r"url\(\s*(['\"]?)([^'\")]+?)\1\s*\)",
        _collapse_unsafe_url,
        stripped,
        flags=re.IGNORECASE,
    )
    if re.search(r"(^|[\s>])<head\b([^>]*)>", stripped, flags=re.IGNORECASE):
        return re.sub(r"(^|[\s>])<head\b([^>]*)>", rf"\1<head\2>{CSP_META}", stripped, count=1, flags=re.IGNORECASE)
    return CSP_META + stripped

File: shared/html-sanitizer.js
const GOOGLE_FONT_HOST_RE = /^https:\/\/fonts\.(googleapis|gstatic)\.com\//i;

const CSP_META =
  "<meta http-equiv=\"Content-Security-Policy\" content=\"default-src 'none'; style-src 'unsafe-inline' https://fonts.googleapis.com; img-src data: file:; font-src data: https://fonts.gstatic.com;\">";

const PREVIEW_CSP_META =
  "<meta http-equiv=\"Content-Security-Policy\" content=\"default-src 'none'; style-src 'unsafe-inline'; img-src data: blob:; font-src data: blob:; media-src data: blob:; connect-src 'none'; script-src 'none'; object-src 'none'; base-uri 'none'; form-action 'none'\">";

const CANVAS_PREVIEW_CSP_META =
  "<meta http-equiv=\"Content-Security-Policy\" content=\"default-src 'none'; style-src 'unsafe-inline' https://fonts.googleapis.com; img-src data: blob:; font-src data: https://fonts.gstatic.com; connect-src 'none'; script-src 'none'; object-src 'none'; base-uri 'none'; form-action 'none'\">";

const SAFE_DATA_URI_PREFIXES = [
  'data:image/png',
  'data:image/jpeg',
  'data:image/jpg',
  'data:image/gif',
  'data:image/bmp',
  'data:image/webp',
  'data:image/x-icon',
];

function isSafeDataUrl(url) {
  const lowered = String(url).trim().toLowerCase();
  return SAFE_DATA_URI_PREFIXES.some((prefix) => lowered.startsWith(prefix));
}

function isAllowedGoogleFontUrl(url) {
  return GOOGLE_FONT_HOST_RE.test(String(url).trim());
}

// Numeric refs decode in browsers without the trailing semicolon; named refs
// keep it required (the no-semicolon legacy named set is not decoded inside
// attributes when followed by alnum/'=', so requiring ';' matches that).
const _URL_ENTITY_RE = /&(?:#(x[0-9a-fA-F]+|\d+);?|([a-zA-Z][a-zA-Z0-9]*);)/g;
const _NAMED_URL_ENTITIES = {
  colon: ':', tab: '\t', newline: '\n', sol: '/', bsol: '\\',
  period: '.', amp: '&', lt: '<', gt: '>', quot: '"', apos: "'",
};

function _decodeUrlEntities(value) {
  return String(value).replace(_URL_ENTITY_RE, (m, num, name) => {
    if (num !== undefined) {
      const code = num[0].toLowerCase() === 'x' ? parseInt(num.slice(1), 16) : parseInt(num, 10);
      return Number.isInteger(code) && code >= 0 && code <= 0x10ffff ? String.fromCodePoint(code) : m;
    }
    const mapped = _NAMED_URL_ENTITIES[name.toLowerCase()];
    return mapped !== undefined ? mapped : m;
  });
}

const URL_ATTR_RE = /(href|src|xlink:href|srcset|poster|action|formaction|background|cite|ping|longdesc)\s*=\s*(['"]?)\s*([^"'>]+)\2/gi;

function neutralizeUrlAttr(match, attr, quote, urlValue) {
  const decoded = _decodeUrlEntities(urlValue);
  const cleaned = decoded.replace(/\s+/g, '').toLowerCase();
  const schemeMatch = cleaned.match(/^([a-z][a-z0-9+.-]*):/);
  const scheme = schemeMatch ? schemeMatch[1] : '';
  if (cleaned.startsWith('data:')) {
    if (!isSafeDataUrl(decoded)) return `${attr}=${quote}${quote}`;
    if (cleaned.startsWith('data:text/html')) return `${attr}=${quote}${quote}`;
    return match;
  }
  if (isAllowedGoogleFontUrl(decoded)) {
    return match;
  }
  if (scheme === 'javascript' || scheme === 'vbscript') {
    return `${attr}=${quote}${quote}`;
  }
  if (scheme === 'http' || scheme === 'https' || scheme === 'file') {
    return `${attr}=${quote}${quote}`;
  }
  return match;
}

function neutralizeSrcsetAttr(match, attr, quote, urlValue) {
  const decoded = _decodeUrlEntities(urlValue);
  for (const candidate of decoded.split(',')) {
    const url = candidate.trim().split(/\s+/)[0] || '';
    const cleaned = url.replace(/\s+/g, '').toLowerCase();
    if (cleaned.startsWith('data:')) {
      if (!isSafeDataUrl(url) || cleaned.startsWith('data:text/html')) {
        return `${attr}=${quote}${quote}`;
      }
      continue;
    }
    const schemeMatch = cleaned.match(/^([a-z][a-z0-9+.-]*):/);
    const scheme = schemeMatch ? schemeMatch[1] : '';
    if (scheme === 'javascript' || scheme === 'vbscript' || scheme === 'http' || scheme === 'https' || scheme === 'file') {
      return `${attr}=${quote}${quote}`;
    }
  }
  return match;
}

function dispatchUrlAttr(match, attr, quote, urlValue) {
  if (String(attr).toLowerCase() === 'srcset') {
    return neutralizeSrcsetAttr(match, attr, quote, urlValue);
  }
  return neutralizeUrlAttr(match, attr, quote, urlValue);
}

function neutralizeImportStatement(match, quote, urlValue) {
  const decoded = _decodeUrlEntities(urlValue);
  if (isSafeDataUrl(decoded)) return match;
  if (isAllowedGoogleFontUrl(decoded)) return match;
  return '';
}

function stripOrKeepLink(fullTag) {
  const hrefMatch = fullTag.match(/\bhref\s*=\s*(['"])([^'"]+)\1/i)
    || fullTag.match(/\bhref\s*=\s*([^\s>]+)/i);
  if (!hrefMatch) return '';
  const href = hrefMatch[2] || hrefMatch[1];
  return isAllowedGoogleFontUrl(href) ? fullTag : '';
}

function sanitizeHtmlForPdf(html) {
  const stripped = String(html)
    .replace(/<!--[\s\S]*?-->/g, '')
    .replace(/<meta[^>]+http-equiv=["']?Content-Security-Policy["']?[^>]*>/gi, '')
    .replace(/<meta[^>]*http-equiv[^>]*>/gi, '')
    .replace(/<base[^>]*>/gi, '')
    .replace(/<\/base>/gi, '')
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<iframe[^>]*>[\s\S]*?<\/iframe>/gi, '')
    .replace(/<object[^>]*>[\s\S]*?<\/object>/gi, '')
    .replace(/<embed[^>]*>/gi, '')
    .replace(/<link[^>]*>/gi, (tag) => stripOrKeepLink(tag))
    .replace(/<script[^>]*>/gi, '')
    .replace(/<\/script>/gi, '')
    .replace(/<iframe[^>]*>/gi, '')
    .replace(/<\/iframe>/gi, '')
    .replace(/\son[a-z]+\s*=\s*"[^"]*"/gi, '')
    .replace(/\son[a-z]+\s*=\s*'[^']*'/gi, '')
    .replace(/\son[a-z]+\s*=\s*`[^`]*`/gi, '')
    .replace(/\son[a-z]+\s*=\s*[^\s>]+/gi, '')
    .replace(/\son[a-z]+\b(?=\s|>|\/)/gi, '')
    .replace(URL_ATTR_RE, dispatchUrlAttr)
    .replace(/@import\s+(['"])([^'"]*)\1\s*;?/gi, neutralizeImportStatement)
    .replace(/expression\s*\(/gi, '')
    .replace(/url\(\s*(['"]?)\s*(?:javascript|vbscript):[^'")\s]*\1\s*\)/gi, "url('')")
    .replace(/url\(\s*(['"]?)([^'")]+?)\1\s*\)/gi, (match, _quote, urlValue) => {
      const decoded = _decodeUrlEntities(urlValue);
      if (isSafeDataUrl(decoded)) return match;
      if (isAllowedGoogleFontUrl(decoded)) return match;
      return "url('')";
    });
  if (/(^|[\s>])<head\b([^>]*)>/i.test(stripped)) {
    return stripped.replace(/(^|[\s>])<head\b([^>]*)>/i, `$1<head$2>${CSP_META}`);
  }
  return `${CSP_META}${stripped}`;
}

function sanitizeHtmlForPreview(html) {
  const raw = String(html);
  let stripped = raw
    .replace(/<!--[\s\S]*?-->/g, '')
    .replace(/<meta[^>]+http-equiv=["']?Content-Security-Policy["']?[^>]*>/gi, '')
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<script[^>]*>/gi, '')
    .replace(/<\/script>/gi, '')
    .replace(/<iframe[^>]*>[\s\S]*?<\/iframe>/gi, '')
    .replace(/<iframe[^>]*>/gi, '')
    .replace(/<\/iframe>/gi, '')
    .replace(/<object[^>]*>[\s\S]*?<\/object>/gi, '')
    .replace(/<object[^>]*>/gi, '')
    .replace(/<\/object>/gi, '')
    .replace(/<embed[^>]*>/gi, '')
    .replace(/<\/embed>/gi, '')
    .replace(/<link[^>]*>/gi, '')
    .replace(/<base[^>]*>/gi, '')
    .replace(/<\/base>/gi, '')
    .replace(/<meta[^>]*http-equiv[^>]*>/gi, '')
    .replace(/\son[a-z]+\s*=\s*"[^"]*"/gi, '')
    .replace(/\son[a-z]+\s*=\s*'[^']*'/gi, '')
    .replace(/\son[a-z]+\s*=\s*`[^`]*`/gi, '')
    .replace(/\son[a-z]+\s*=\s*[^\s>]+/gi, '')
    .replace(/\son[a-z]+\b(?=\s|>|\/)/gi, '')
    .replace(URL_ATTR_RE, (match, attr, quote, urlValue) => {
      if (String(attr).toLowerCase() === 'srcset') {
        return neutralizeSrcsetAttr(match, attr, quote, urlValue);
      }
      const decoded = _decodeUrlEntities(urlValue);
      const lowered = decoded.trim().toLowerCase();
      const q = quote || '"';
      if (lowered.startsWith('data:')) {
        return isSafeDataUrl(decoded) ? match : `${attr}=${q}${q}`;
      }
      if (lowered.startsWith('blob:')) return match;
      return `${attr}=${q}${q}`;
    })
    .replace(/url\(\s*(['"]?)\s*(?:javascript|vbscript):[^'")\s]*\1\s*\)/gi, "url('')")
    .replace(/url\(\s*(['"]?)([^'")]+?)\1\s*\)/gi, (match, _quote, urlValue) => {
      const decoded = _decodeUrlEntities(urlValue);
      const lowered = decoded.trim().toLowerCase();
      if (lowered.startsWith('blob:')) return match;
      if (lowered.startsWith('data:')) {
        return isSafeDataUrl(decoded) ? match : "url('')";
      }
      return "url('')";
    })
    .replace(/@import\s+[^;]+;/gi, '')
    .replace(/expression\s*\(/gi, '');

  stripped = stripped.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, (block) => {
    let inner = block.replace(/^<style[^>]*>/i, '').replace(/<\/style>$/i, '');
    inner = inner.replace(/url\([^)]+\)/gi, (m) => {
      const urlMatch = m.match(/url\(\s*(['"]?)([^'")]+)\1\s*\)/i);
      if (!urlMatch) return "url('')";
      const v = String(urlMatch[2]).trim().toLowerCase();
      if (v.startsWith('blob:')) return m;
      if (v.startsWith('data:') && isSafeDataUrl(urlMatch[2])) return m;
      return "url('')";
    });
    inner = inner.replace(/@import[^;]+;/gi, '');
    return `<style>${inner}</style>`;
  });

  if (/(^|[\s>])<head\b([^>]*)>/i.test(stripped)) {
    return stripped.replace(/(^|[\s>])<head\b([^>]*)>/i, `$1<head$2>${PREVIEW_CSP_META}`);
  }
  return `${PREVIEW_CSP_META}${stripped}`;
}

function sanitizeHtmlForCanvasPreview(html) {
  const stripped = String(html)
    .replace(/<!--[\s\S]*?-->/g, '')
    .replace(/<meta[^>]+http-equiv=["']?Content-Security-Policy["']?[^>]*>/gi, '')
    .replace(/<meta[^>]*http-equiv[^>]*>/gi, '')
    .replace(/<base[^>]*>/gi, '')
    .replace(/<\/base>/gi, '')
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<iframe[^>]*>[\s\S]*?<\/iframe>/gi, '')
    .replace(/<object[^>]*>[\s\S]*?<\/object>/gi, '')
    .replace(/<embed[^>]*>/gi, '')
    .replace(/<link[^>]*>/gi, (tag) => stripOrKeepLink(tag))
    .replace(/<script[^>]*>/gi, '')
    .replace(/<\/script>/gi, '')
    .replace(/<iframe[^>]*>/gi, '')
    .replace(/<\/iframe>/gi, '')
    .replace(/\son[a-z]+\s*=\s*"[^"]*"/gi, '')
    .replace(/\son[a-z]+\s*=\s*'[^']*'/gi, '')
    .replace(/\son[a-z]+\s*=\s*`[^`]*`/gi, '')
    .replace(/\son[a-z]+\s*=\s*[^\s>]+/gi, '')
    .replace(/\son[a-z]+\b(?=\s|>|\/)/gi, '')
    .replace(URL_ATTR_RE, dispatchUrlAttr)
    .replace(/@import\s+(['"])([^'"]*)\1\s*;?/gi, neutralizeImportStatement)
    .replace(/expression\s*\(/gi, '')
    .replace(/url\(\s*(['"]?)\s*(?:javascript|vbscript):[^'")\s]*\1\s*\)/gi, "url('')")
    .replace(/url\(\s*(['"]?)([^'")]+?)\1\s*\)/gi, (match, _quote, urlValue) => {
      const decoded = _decodeUrlEntities(urlValue);
      const lowered = decoded.trim().toLowerCase();
      if (lowered.startsWith('blob:')) return match;
      if (isSafeDataUrl(decoded)) return match;
      if (isAllowedGoogleFontUrl(decoded)) return match;
      return "url('')";
    });

  if (/(^|[\s>])<head\b([^>]*)>/i.test(stripped)) {
    return stripped.replace(/(^|[\s>])<head\b([^>]*)>/i, `$1<head$2>${CANVAS_PREVIEW_CSP_META}`);
  }
  return `${CANVAS_PREVIEW_CSP_META}${stripped}`;
}

module.exports = { sanitizeHtmlForPdf, sanitizeHtmlForPreview, sanitizeHtmlForCanvasPreview, CSP_META, PREVIEW_CSP_META, CANVAS_PREVIEW_CSP_META, isSafeDataUrl, isAllowedGoogleFontUrl };
